File: internal/worddiff/worddiff.go
// Package worddiff computes a token-level ("word") diff between two short
// strings using a classic longest-common-subsequence (LCS) dynamic program.
// It is meant for refining a single changed line into intra-line highlights:
// given the old and new text of a replaced line, it returns the runs of text
// that stayed the same and the runs that changed, on each side.
//
// Ported from ayame-editor's web/src/search.ts inlineWordDiff (see
// hjosugi/ayame-diff#8). The LCS DP is O(m*n) in time and memory, so callers
// must keep inputs small; the MaxChars/MaxTokens guards below enforce that.
package worddiff

import (
	"regexp"
	"strings"
	"unicode"
	"unicode/utf8"
)

// Segment is a run of contiguous text on one side of the diff. Changed is true
// when the run has no counterpart on the other side (an insertion or deletion),
// false when it is common to both sides.
type Segment struct {
	Text    string
	Changed bool
}

// Result holds the segmented old and new text. Concatenating the Text of Old
// reproduces oldText; concatenating New reproduces newText. Common (unchanged)
// segments appear in both Old and New in the same order.
type Result struct {
	Old []Segment
	New []Segment
}

const (
	// MaxChars bounds the combined input size. Beyond this the O(m*n) DP is
	// not worth running for an inline highlight, so Diff bails out.
	MaxChars = 2000
	// MaxTokens bounds the combined token count, the real driver of the DP's
	// m*n cost.
	MaxTokens = 260
)

// tokenRE splits text into coarse runs of (1) whitespace, (2) Unicode word
// characters — letters, combining marks, numbers, or underscore — or (3) any
// other characters. Including combining marks (\p{M}) in the word class keeps a
// base letter and its mark together (e.g. "e"+U+0301 stays one token) instead
// of stranding the mark as its own "other" run. CJK word runs are broken down
// further by Tokenize. Compiled once here to match the reference's single
// shared RegExp.
var tokenRE = regexp.MustCompile(`(\s+|[\p{L}\p{M}\p{N}_]+|[^\s\p{L}\p{M}\p{N}_]+)`)

// cjkScripts lists the scripts written without spaces between words. A whole
// run of them would otherwise be a single token, so the inline diff could only
// mark the entire run changed — useless for Japanese, where "日本語" vs "日本国"
// should highlight just the last character. Splitting each such character into
// its own token lets the diff align character by character.
var cjkScripts = []*unicode.RangeTable{
	unicode.Han, unicode.Hiragana, unicode.Katakana, unicode.Hangul,
}

func isCJK(r rune) bool { return unicode.IsOneOf(cjkScripts, r) }

// Tokenize splits s into diff tokens. An empty string yields no tokens. Word
// runs are emitted whole, except that each CJK character becomes its own token
// so spaceless CJK text aligns per character rather than as one block.
func Tokenize(s string) []string {
	coarse := tokenRE.FindAllString(s, -1)
	tokens := make([]string, 0, len(coarse))
	for _, tok := range coarse {
		tokens = appendSplitCJK(tokens, tok)
	}
	return tokens
}

// appendSplitCJK appends tok to out, emitting every CJK character as its own
// token while keeping maximal non-CJK runs intact. Tokens with no CJK character
// (the common case — ASCII words, numbers, punctuation) take a fast path and
// are appended unchanged.
func appendSplitCJK(out []string, tok string) []string {
	hasCJK := false
	for _, r := range tok {
		if isCJK(r) {
			hasCJK = true
			break
		}
	}
	if !hasCJK {
		return append(out, tok)
	}
	var buf strings.Builder
	for _, r := range tok {
		if isCJK(r) {
			if buf.Len() > 0 {
				out = append(out, buf.String())
				buf.Reset()
			}
			out = append(out, string(r))
		} else {
			buf.WriteRune(r)
		}
	}
	if buf.Len() > 0 {
		out = append(out, buf.String())
	}
	return out
}

// Diff returns the word-level diff of oldText vs newText, or (nil, false) when
// they are equal or exceed the guard limits.
//
// The size guard uses utf8.RuneCountInString rather than len (byte length).
// The reference implementation in TypeScript compares against String.length,
// which counts UTF-16 code units; counting runes here is the closest
// Unicode-correct analogue and keeps the effective limit consistent regardless
// of how many bytes a character encodes to.
func Diff(oldText, newText string) (*Result, bool) {
	if oldText == newText {
		return nil, false
	}
	if utf8.RuneCountInString(oldText)+utf8.RuneCountInString(newText) > MaxChars {
		return nil, false
	}

	oldTokens := Tokenize(oldText)
	newTokens := Tokenize(newText)
	if len(oldTokens)+len(newTokens) > MaxTokens {
		return nil, false
	}

	m, n := len(oldTokens), len(newTokens)

	// dp[i][j] = length of the LCS of oldTokens[i:] and newTokens[j:]. Filled
	// backward so the forward walk below can read the "best from here" value.
	dp := make([][]int, m+1)
	for i := range dp {
		dp[i] = make([]int, n+1)
	}
	for i := m - 1; i >= 0; i-- {
		for j := n - 1; j >= 0; j-- {
			if oldTokens[i] == newTokens[j] {
				dp[i][j] = dp[i+1][j+1] + 1
			} else if dp[i+1][j] >= dp[i][j+1] {
				dp[i][j] = dp[i+1][j]
			} else {
				dp[i][j] = dp[i][j+1]
			}
		}
	}

	res := &Result{}
	i, j := 0, 0
	for i < m || j < n {
		switch {
		case i < m && j < n && oldTokens[i] == newTokens[j]:
			// Common token: emit unchanged on both sides.
			res.Old = pushDiffPart(res.Old, oldTokens[i], false)
			res.New = pushDiffPart(res.New, newTokens[j], false)
			i++
			j++
		case j >= n || (i < m && dp[i+1][j] >= dp[i][j+1]):
			// Advancing in old loses no more of the LCS than advancing in new
			// (ties favor consuming old): this old token was deleted.
			res.Old = pushDiffPart(res.Old, oldTokens[i], true)
			i++
		default:
			// Otherwise this new token was inserted.
			res.New = pushDiffPart(res.New, newTokens[j], true)
			j++
		}
	}
	return res, true
}

// pushDiffPart appends text to parts, merging it into the last segment when the
// Changed flag matches so adjacent runs of the same kind form one segment.
// Empty text is never pushed. Mirrors the reference's pushDiffPart.
func pushDiffPart(parts []Segment, text string, changed bool) []Segment {
	if text == "" {
		return parts
	}
	if n := len(parts); n > 0 && parts[n-1].Changed == changed {
		parts[n-1].Text += text
		return parts
	}
	return append(parts, Segment{Text: text, Changed: changed})
}

File: internal/server/web/worddiff.js
// Word-level diff used by the side-by-side view.
//
// It lives outside app.js so it can be executed in tests without a DOM: it is
// the one piece of client logic that is a real algorithm rather than DOM
// plumbing, and it had no execution-level coverage at all (#139). The wrapper
// mirrors syntax.js — a browser global plus a CommonJS export for node --test.
(function (root) {
  "use strict";

  const INLINE_MAX_CHARS = 2000;
  const INLINE_MAX_TOKENS = 260;
  // Word class includes combining marks (\p{Mark}) so a base letter keeps its
  // accent (e.g. "e"+U+0301) instead of stranding the mark as a separate token.
  const TOKEN_RE = /(\s+|[\p{Letter}\p{Mark}\p{Number}_]+|[^\s\p{Letter}\p{Mark}\p{Number}_]+)/gu;
  // CJK is written without spaces, so a whole run would be one token and the diff
  // could only mark it all changed; split each such character out so "日本語" vs
  // "日本国" highlights just the last character. Mirrors worddiff.go's cjkScripts.
  const CJK_RE = /[\p{Script=Han}\p{Script=Hiragana}\p{Script=Katakana}\p{Script=Hangul}]/u;

  function inlineTokens(text) {
    const tokens = [];
    for (const m of String(text || "").matchAll(TOKEN_RE)) appendSplitCJK(tokens, m[0]);
    return tokens;
  }
  // appendSplitCJK pushes tok, emitting each CJK character as its own token while
  // keeping maximal non-CJK runs intact. Non-CJK tokens take a fast path.
  function appendSplitCJK(tokens, tok) {
    if (!CJK_RE.test(tok)) { tokens.push(tok); return; }
    let buf = "";
    for (const ch of tok) { // iterates by code point
      if (CJK_RE.test(ch)) {
        if (buf) { tokens.push(buf); buf = ""; }
        tokens.push(ch);
      } else {
        buf += ch;
      }
    }
    if (buf) tokens.push(buf);
  }
  function pushPart(parts, text, changed) {
    if (!text) return;
    const last = parts[parts.length - 1];
    if (last && last.changed === changed) last.text += text;
    else parts.push({ text, changed });
  }
  // inlineDPScratch is reused across calls so a full render does not allocate a
  // fresh table per changed line. It is cleared to the needed length each time,
  // since the algorithm reads cells it has not written yet (row m and column n
  // must be zero).
  let inlineDPScratch = new Uint16Array(0);

  function inlineDPBuffer(size) {
    if (inlineDPScratch.length < size) inlineDPScratch = new Uint16Array(size);
    inlineDPScratch.fill(0, 0, size);
    return inlineDPScratch;
  }

  function inlineWordDiff(oldText, newText) {
    oldText = String(oldText || "");
    newText = String(newText || "");
    if (oldText === newText) return null;
    if (oldText.length + newText.length > INLINE_MAX_CHARS) return null;
    const a = inlineTokens(oldText);
    const b = inlineTokens(newText);
    if (a.length + b.length > INLINE_MAX_TOKENS) return null;
    const m = a.length, n = b.length;
    // One flat Uint16Array rather than m+1 of them: the row-per-line form
    // allocated an array per token and turned every dp read into an
    // array-of-arrays double dereference. The table is bounded by
    // INLINE_MAX_TOKENS above, so a single buffer is reused across calls (#155).
    const stride = n + 1;
    const dp = inlineDPBuffer(stride * (m + 1));
    const at = (i, j) => i * stride + j;
    for (let i = m - 1; i >= 0; i--)
      for (let j = n - 1; j >= 0; j--)
        dp[at(i, j)] = a[i] === b[j] ? dp[at(i + 1, j + 1)] + 1 : Math.max(dp[at(i + 1, j)], dp[at(i, j + 1)]);
    const oldParts = [], newParts = [];
    let i = 0, j = 0;
    while (i < m || j < n) {
      if (i < m && j < n && a[i] === b[j]) {
        pushPart(oldParts, a[i], false);
        pushPart(newParts, b[j], false);
        i++; j++;
      } else if (j >= n || (i < m && dp[at(i + 1, j)] >= dp[at(i, j + 1)])) {
        pushPart(oldParts, a[i], true); i++;
      } else {
        pushPart(newParts, b[j], true); j++;
      }
    }
    return { oldParts, newParts };
  }

  const api = { inlineWordDiff, inlineTokens, pushPart, INLINE_MAX_CHARS, INLINE_MAX_TOKENS };
  root.AyameWordDiff = api;
  if (typeof module !== "undefined" && module.exports) module.exports = api;
})(typeof globalThis !== "undefined" ? globalThis : window);
